### PR TITLE
Array operator

### PR DESCRIPTION
**Pull request by matthias-springer**

### CODE REVIEW

**Comment by yzyzsun** [comment]
I have just resolved the conflicts with `master`.

File: ikra/soa/array_operator.h [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+           int AddressMode2, int StorageMode2, typename Owner2>
+#define AOS_ARRAY_FIELD(n) AosArrayField_<std::array<T, ArraySize>, \
+  Capacity ## n, Offset ## n, AddressMode ## n, StorageMode ## n, Owner ## n>
+#define SOA_ARRAY_FIELD(n) SoaArrayField_<T, ArraySize, \
+  Capacity ## n, Offset ## n, AddressMode ## n, StorageMode ## n, Owner ## n>
+#define STD_ARRAY std::array<T, ArraySize>
+
+#define IKRA_DEFINE_STD_ARRAY_OPERATORS_A(symbol) \
+  template<typename T, size_t N> \
+  std::array<T, N> operator symbol (std::array<T, N> left, \
```
Comment (matthias-springer) [inline ikra/soa/array_operator.h]: Provide overload where `left` and `right` are passed as reference. Put in separate commit/pull request. (Put TODO comment for now.)

File: ikra/soa/array_operator.h [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+                             AOS_ARRAY_FIELD(2)& right) { \
+    (STD_ARRAY&)left symbol ## = (STD_ARRAY&)right; \
+  } \
+  TEMPLATE_ARRAY_FIELD_W \
+  void operator symbol ## = (AOS_ARRAY_FIELD(1)& left, \
+                             SOA_ARRAY_FIELD(2)& right) { \
+    (STD_ARRAY&)left symbol ## = right; \
+  }
+
+#define IKRA_DEFINE_AOS_ARRAY_FIELD_OPERATORS_M(symbol) \
```
Comment (matthias-springer) [inline ikra/soa/array_operator.h]: Rename to `_SCALAR`. Provide both scalar and vector operations for all operators.

File: ikra/soa/array_operator.h [hunk 3]
```diff
@@ -0,0 +1,260 @@
+#ifndef SOA_ARRAY_OPERATOR_H
+#define SOA_ARRAY_OPERATOR_H
+
+#include "soa/array_field.h"
+#include "soa/util.h"
+
+// This is used for operations between std::arrays and array fields
```
Comment (yzyzsun) [inline ikra/soa/array_operator.h]: Put `.` at the end.

File: ikra/soa/array_operator.h [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+// For `IKRA_TEMPLATE_ARRAY_FIELD`, just write `XXX_ARRAY_FIELD()`;
+// For `IKRA_TEMPLATE_ARRAY_FIELD_W`, write like `XXX_ARRAY_FIELD(1)`.
+#define IKRA_AOS_ARRAY_FIELD(n) AosArrayField_<std::array<T, ArraySize>, \
+  Capacity ## n, Offset ## n, AddressMode ## n, StorageMode ## n, Owner ## n>
+#define IKRA_SOA_ARRAY_FIELD(n) SoaArrayField_<T, ArraySize, \
+  Capacity ## n, Offset ## n, AddressMode ## n, StorageMode ## n, Owner ## n>
+#define IKRA_STD_ARRAY std::array<T, ArraySize>
+
+// TODO: Provide overload where `left` and `right` are passed as reference
+#define IKRA_DEFINE_IKRA_STD_ARRAY_OPERATORS_VECTOR(symbol) \
```
Comment (yzyzsun) [inline ikra/soa/array_operator.h]: Comment like "Given a symbol, defines an operator between all combinations of ...".

File: ikra/soa/array_operator.h [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+  } \
+  IKRA_TEMPLATE_ARRAY_FIELD \
+  void operator symbol ## = (IKRA_STD_ARRAY& left, \
+                             IKRA_SOA_ARRAY_FIELD()& right) { \
+    for (size_t i = 0; i < ArraySize; ++i) { \
+      left[i] symbol ## = right[i]; \
+    } \
+  }
+
+#define IKRA_DEFINE_IKRA_STD_ARRAY_OPERATORS_SCALAR(symbol) \
```
Comment (yzyzsun) [inline ikra/soa/array_operator.h]: Comments.
